File: Data Structures & Algorithms/reorder-linked-list/submission-0.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution {
    public void ReorderList(ListNode head) {
        List<int> arr = [];
        ListNode current = head;
        while (current != null) {
            arr.Add(current.val);
            current = current.next;
        }

        if(arr.Count<3) {return;}
        if(arr.Count == 3) {
            head.next.next.next = head.next;
            head.next = head.next.next;
            head.next.next.next = null;
            return;
        }
        
        int addSize = arr.Count/2;
        
        current = head;

        int last = arr.Count-1;
        while(addSize>0) {
            //Console.WriteLine("Running");
            ListNode h1 = current.next;
            ListNode node =  new ListNode(arr[last], h1);
            current.next = node;
            addSize--;
            last--;
            if(addSize == 0) {
                current = current.next;
                if(arr.Count%2 == 1) {
                    current = current.next;
                }
                current.next = null;
            }
            else{
                current = current.next.next;
            }
        }

    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int count = 0;
        ListNode* current = head;
        while(current) {
            count++;
            current = current->next;
        }

        if(n==count) {
            head = head->next;
            return head;
        }

        count -= ++n;

        current = head;
        
        while(count) {
            current = current->next;
            count--;
        }
        
        ListNode* hold = current->next;
        current->next = nullptr;

        if(hold->next) {
            current->next = hold->next;
        }

        return head;
    }
};
